File: Backend/gamificationService/src/main/java/com/campusConnect/gamificationService/service/GamificationService.java
package com.campusConnect.gamificationService.service;

import com.campusConnect.gamificationService.dto.GamificationDTO;
import com.campusConnect.gamificationService.dto.LeaderBoardDTO;

import java.util.List;

public interface GamificationService {
    // Gamification operations
    GamificationDTO addPoints(GamificationDTO gamificationDTO);
    List<GamificationDTO> getUserPoints(Long userId);
    Long getTotalPoints(Long userId);

    // Leaderboard operations
    List<LeaderBoardDTO> getLeaderboard();
    void updateLeaderboard();
}

File: frontend/src/api/gamificationService.ts
import api from './client'

export interface GamificationDTO {
  userId: string
  points: number
  level: number
}

export interface LeaderboardDTO {
  rank: number
  userId: string
  userName: string
  points: number
  level: number
}

export interface BadgeDTO {
  id: string
  name: string
  description: string
  icon?: string
  userId: string
  earnedAt: string
}

export const gamificationService = {
  getUserPoints: (userId: string) =>
    api.get<GamificationDTO>(`/gamification/points/user/${userId}`),

  getTotalPoints: (userId: string) =>
    api.get<{ totalPoints: number }>(`/gamification/points/total/${userId}`),

  addPoints: (userId: string, points: number, reason?: string) =>
    api.post('/gamification/points', { userId, points, reason }),

  getLeaderboard: (semester?: number, limit = 10) =>
    api.get<LeaderboardDTO[]>('/gamification/leaderboard', { params: { semester, limit } }),

  updateLeaderboard: () =>
    api.post('/gamification/leaderboard/update'),

  // Badges
  getUserBadges: (userId: string) =>
    api.get<BadgeDTO[]>(`/gamification/badges/user/${userId}`),

  awardBadge: (userId: string, badgeName: string) =>
    api.post(`/gamification/badge/${userId}`, { badgeName }),
}

